### PR TITLE
process_pending_deposits - transaction processing refactor

### PR DESCRIPTION
**Pull request by stfung77**
added database fields for transaction: queue, queued_for_submit, submission_status
removed references to pending_execution_attempt in process_pending_deposits
removed multisig checking
use asyncio Task/Queues for transaction processing
added heartbeat/lock to ensure only one instance of process_pending_deposits is running

image
image

TODO: add more comments in code
TODO: replace actual process_pending_desposits file
TODO: update unit tests

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ services:
       - ./example/data:/code/data
     depends_on:
       - server
-    command: watchmedo auto-restart --recursive --pattern="*.py;*.env" --directory="." python manage.py process_pending_deposits -- --loop
+    command: watchmedo auto-restart --recursive --pattern="*.py;*.env" --directory="." python manage.py test_process_pending_deposits
```
Comment (stfung77) [inline docker-compose.yml:44]: the test_ will be removed

File: polaris/polaris/management/commands/process_pending_deposits.py [hunk 2]
```diff
@@ -9,23 +9,14 @@
 
 import django.db.transaction
 from django.core.management import BaseCommand
-from stellar_sdk import (
```
Comment (stfung77) [inline polaris/polaris/management/commands/process_pending_deposits.py:12]: this file will be replaced with the changes in test_process_pending_deposits

File: polaris/polaris/management/commands/test_process_pending_deposits.py [hunk 3]
```diff
@@ -0,0 +1,928 @@
...
+        transaction.status_message = message
+        transaction.status = Transaction.STATUS.error
+        transaction.save()
+        logger.error(message)
+
+    @classmethod
+    def handle_transaction_submission_exception(cls, transaction, exception, message):
+        if type(exception) is TransactionSubmissionBlocked:
+            transaction.submission_status = Transaction.SUBMISSION_STATUS.blocked
+        elif type(exception) is TransactionSubmissionFailed:
```
Comment (stfung77) [inline polaris/polaris/management/commands/test_process_pending_deposits.py:812]: TODO: add a transaction.submission_status.failed

**Review by JakeUrban — COMMENTED** [review]
I ran through a deposit flow with this branch for an account that existed and had a trustline and it worked 🎉 . However, if it didn't have a trustline or didn't exist, it ran into some issues, which I'm pretty sure are just related to calling `sync_to_async()` properly. I left a couple comments explaining.

I haven't been able to really deep dive into the code yet, but made a couple comments about documenting the changes to the data model. I think we still need to remove `pending_execution_attempt` as well.

File: polaris/polaris/management/commands/test_process_pending_deposits.py [hunk 4]
```diff
@@ -0,0 +1,928 @@
...
+                    logger.debug(
+                        f"locked to create destination account for transaction {transaction.id}"
+                    )
+                while True:
+                    try:
+                        rci.create_destination_account(transaction=transaction)
+                    except TransactionSubmissionPending as e:
+                        await sync_to_async(
+                            ProcessPendingDeposits.handle_transaction_submission_exception()
+                            )(transaction, e, "")
```
[the remark this answers is not in the document]
Comment (stfung77) [reply polaris/polaris/management/commands/test_process_pending_deposits.py]: updated. i'll also add some unit tests to hit these code paths

[the remark this answers is not in the document]
Comment (stfung77) [reply polaris/polaris/management/commands/test_process_pending_deposits.py]: lets remove Transaction.pending_execution_attempt in a separate PR. i see some references to it in other files as well.

**Comment by JakeUrban** [comment]
@stfung77 just ran the code again and it almost worked! But this time it was my code with the bug :-)

The `SelfCustodyIntegration` subclass that is the default `CustodyIntegration` classed called from the deposit processing logic has a function called `submit_deposit_transaction()`. This function didn't account for the possibility that the `Transaction.envelope_xdr` attribute could contain a transaction used to create the account instead of a transaction that could be used to send the deposit amount.

So when I used the demo wallet with an account that didn't exist, Polaris created the account, waited for the trustline to be established, and then queued the transaction for submission again. When it called `rci.submit_deposit_transaction()`, Polaris resubmitted the create account transaction, which succeeded (because it had already been submitted) and completed the transaction. So I never got my deposit.

Can adjust the `SelfCustody.submit_deposit_transaction()` function to account for this case? Or, should Polaris clear `Transaction.envelope_xdr`, `Transaction.stellar_transaction_id`, etc. before queueing the transaction for submission again?

**Comment by stfung77** [comment]
> @stfung77 just ran the code again and it almost worked! But this time it was my code with the bug :-)
> 
> The `SelfCustodyIntegration` subclass that is the default `CustodyIntegration` classed called from the deposit processing logic has a function called `submit_deposit_transaction()`. This function didn't account for the possibility that the `Transaction.envelope_xdr` attribute could contain a transaction used to create the account instead of a transaction that could be used to send the deposit amount.
> 
> So when I used the demo wallet with an account that didn't exist, Polaris created the account, waited for the trustline to be established, and then queued the transaction for submission again. When it called `rci.submit_deposit_transaction()`, Polaris resubmitted the create account transaction, which succeeded (because it had already been submitted) and completed the transaction. So I never got my deposit.
> 
> Can you adjust the `SelfCustody.submit_deposit_transaction()` function to account for this case? Or, should Polaris clear `Transaction.envelope_xdr`, `Transaction.stellar_transaction_id`, etc. before queueing the transaction for submission again?

I think Transaction.envelope_xdr should be cleared. i'll take care of this

**Comment by JakeUrban** [comment]
@stfung77 

> I think Transaction.envelope_xdr should be cleared. i'll take care of this

I think thats a good approach, just make sure to wipe all related state as well (like the stellar tx ID). 

In the future, I'd like to use to persist all transactions submitted to the network for a particular `Transaction` object. This will require some schema changes but something to note here.

**Comment by stfung77** [comment]
> @stfung77
> 
> > I think Transaction.envelope_xdr should be cleared. i'll take care of this
> 
> I think thats a good approach, just make sure to wipe all related state as well (like the stellar tx ID).
> 
> In the future, I'd like to persist all transactions submitted to the network for a particular `Transaction` object. This will require some schema changes but something to note here.



> 

Yeah i agree that transactions should be persisted somewhere. What are your thoughts on logging the tx id and envelope? Any security concerns around this?

**Pull request closed without merging**